### PR TITLE
feat(transformers): introduce `matchAlgorithm` option for new matching algorithm

### PR DESCRIPTION
**Pull request by fuma-nama**
### Description

This PR aims to redesign notation transformer utility and its usage to improve performance and make it more intuitive.

#### Changes

- Parse comments separately and cache it into the `data` property of `code` nodes.
- Notation transformers read and replace parsed comments.
- Update original token, remove empty comment tokens and lines if necessary.

#### New Syntax

```js
console.log("hello") // [!code highlight]

console.log("highlighted") // [!code highlight:2]
console.log("highlighted")

// the comment below will be removed
// [!code highlight:2]
console.log("highlighted")
console.log("highlighted")

console.log("combine") // [!code focus] [!code highlight]
console.log("combine") // Comment Allowed [!code focus] [!code highlight]
```

#### JSX

```jsx
<div>
    <p>Hello World</p>
    {/* [!code highlight] */}
    <p>Highlighted</p>
</div>
```

#### Breaking

It only looks for the last few tokens of line to scan comments:

```jsx
console.log("// [!code highlight]"); // will not work
```

It doesn't need multiple `//` to chain notations:

```ts
console.log("test") // [!code highlight] // [!code ++]

// will cause weird result like:
console.log("test") // //
```

#### Edge cases

for usage like:

```ts
// some text [!code highlight:2]
console.log("highlighted")
console.log("highlighted")
```

line 2-3 will be highlighted, but the original comment is kept.

#### Backward Compatibility

Added a `legacy` option to force legacy behaviors:

```ts
transformerNotationFocus({ legacy: true })
transformerNotationHighlight({ legacy: true })
transformerNotationErrorLevel({ legacy: true })
```

### Linked Issues

#619 #770 

### Additional context

this PR will introduce breaking changes, I currently made a separate package to avoid the problems. Backward compatibility is possible (edit: supported), but it takes extra care and further discussion.

code implementations:
- the `createCommentNotationTransformer` parses comment with `parseComments` and cache the result.
- each parsed comment object contains related hast nodes, and `info`:
  ```
      /* hello world */
   |----|-----------|-|
   prefix   content  suffix
  ```
- remove nodes when `content` became empty.

- transformers using `createCommentNotationTransformer` still provides a regex, but the regex no longer need to match the comment

### CODE REVIEW

File: packages/transformers/src/shared/notation-transformer.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      }
+
+      data._shiki_notation ??= parseComments(lines, ['jsx', 'tsx'].includes(this.options.lang))
+      const parsed = data._shiki_notation
+
+      for (const comment of parsed) {
+        if (comment.info[1].length === 0)
+          continue
+
+        const isLineCommentOnly = comment.line.children.length === (comment.isJsxStyle ? 3 : 1)
```
Comment (fuma-nama) [inline packages/transformers/src/shared/notation-transformer.ts:35]: it indicates the line contains only the comment itself, like:
```js
// [!code highlight]
console.log("test")
```
so we will highlight the next line and remove the original line when `content` became empty.

File: packages/transformers/src/transformers/notation-map.ts [hunk 2]
```diff
@@ -24,14 +24,14 @@ export function transformerNotationMap(
 
   return createCommentNotationTransformer(
     name,
-    new RegExp(`\\s*(?://|/\\*|<!--|#|--|%{1,2}|;{1,2}|"|')\\s+\\[!code (${Object.keys(classMap).map(escapeRegExp).join('|')})(:\\d+)?\\]\\s*(?:\\*/|-->)?\\s*$`),
+    new RegExp(`\\s*\\[!code (${Object.keys(classMap).map(escapeRegExp).join('|')})(:\\d+)?\\]`),
```
Comment (fuma-nama) [inline packages/transformers/src/transformers/notation-map.ts:27]: we no longer need to provide the regex for matching comments, it is better given that the name `createCommentNotationTransformer` should this too.

**Comment by fuma-nama** [comment]
I kept the failing unit tests to experiment about some changes for backward compatibility

**Comment by Fro-Q** [comment]
This is what I'm looking for! Are you still working on it ? 'cause I actually need it so bad :>

**Comment by fuma-nama** [comment]
See https://github.com/fuma-nama/shiki-transformers

**Comment by antfu** [comment]
Hey @fuma-nama, thanks for the awesome work and sorry for the long due. In other to process this PR, we have to first make it non-breaking and rolling out it progressively. I have push some commits to introduce the `matchAlgorithm` which will default to the old behaviour, and we will rolling it by default in v3. I also wrote some guide for people to migrate or opt-in, if I messed anything feel free to push and improve it. Thanks
